### PR TITLE
[K12] K12 Warning Message on Education History

### LINKED ISSUE #271: [K12] K12 Warning Message on Education History
**Note: FOR INTERNAL USE ONLY. For support questions, please use the EDA Chatter group in [The Power of Us Hub](https://powerofus.force.com).
For feature requests, please use the Ideas tab in [The Power of Us Hub](https://powerofus.force.com).**

* **Do you want to request a *feature* or report a *bug*?**
Feature

* **What is the current behavior?**
When creating or updating an Education History record with a related Grade Enrollment record, the Education History record is created or updated regardless of whether the specified Educational Institution matches the value assigned to the related Grade Enrollment.

  Similarly, when creating or updating a Grade Enrollment record referencing an Education History record that is not 
  associated with the same Institution specified by the Educational Institution field on the Grade Enrollment, there is no 
  warning message displayed and the Grade Enrollment record is created or updated with the mismatching institutional 
  data.

* **What is the expected behavior?**
When creating or updating an Education History record with a related Grade Enrollment record, an Education History record with a specified Educational Institution that does not match the value assigned to its related Grade Enrollment should display a warning message and the Education History record should not be created or updated.  Blank values for Educational Institution should not be evaluated.

  When creating or updating a Grade Enrollment record that references an Education History record with a 
  specified Educational Institution that does not match the value specified by the Educational Institution field on the Grade 
  Enrollment, a warning message should be displayed and the Grade Enrollment record should not be created or updated.  
  Blank values for Educational Institution should not be evaluated.
 
* **What is the motivation / use case for changing the behavior?**
When creating an Education History record there are multiple scenarios that would allow a user to create bad data. Specifically when the Education Institution on the Education History Record does not match the Education Institution on the related Grade Enrollment. 

* **Acceptance Criteria**

  - **Given**: creating or saving an Education History record
  - **When**: the Education Institution on the Education History Record does not match the Education Institution on the related Grade Enrollment; 
  - **Then**: the user is not allowed to save

### PR DESCRIPTION
**Pull request by m-kinkade**
# Critical Changes

# Changes
We've added TDTM trigger handlers on the Education History and Grade Enrollment objects to ensure data integrity. For more information, see EDUHIS_DataInteg_TDTM.cls and GRER_DataInteg_TDTM.cls in [Apex Class Descriptions](https://powerofus.force.com/s/article/EDA-Apex-Class-Descriptions).

# Issues Closed
Closes #271 

# New Metadata
### Custom Labels
EduHisInstitutionMismatch
GrErInstitutionMismatch

### Apex Classes
EDUHIS_DataInteg_TDTM
EDUHIS_DataInteg_TEST
GRER_DataInteg_TDTM
GRER_DataInteg_TEST

# Deleted Metadata

# Testing Notes
https://salesforce.quip.com/scHGARrggA1N#QZZACAsQxHD

### CODE REVIEW

**Review by juliesnow — APPROVED** [review]
added a link to product docs for more info

File: src/classes/GRER_DataInteg_TEST.cls [hunk 1]
```diff
@@ -0,0 +1,1638 @@
...
+            hed__Education_History__c testEduHist = new hed__Education_History__c(hed__Contact__c = testContact.Id,
+                                                                                  hed__Account__c = null);
+
+            eduHistList.add(testEduHist);
+        }
+
+        insert eduHistList;
+
+        for (hed__Education_History__c eduHist : eduHistList){
+            Grade_Enrollment__c gradeEnroll = new Grade_Enrollment__c(Contact__c = testContact.Id,
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:604]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 2]
```diff
@@ -0,0 +1,1638 @@
...
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:666]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 3]
```diff
@@ -0,0 +1,1638 @@
...
+        for (Integer i = 0; i < 5; i++){
+            hed__Education_History__c testEduHist = new hed__Education_History__c(hed__Contact__c = testContact.Id,
+                                                                                  hed__Account__c = firstAccount.Id);
+            eduHistList.add(testEduHist);
+        }
+
+        insert eduHistList;
+
+        for (hed__Education_History__c eduHist : eduHistList){
+            Grade_Enrollment__c gradeEnroll = new Grade_Enrollment__c(Contact__c = testContact.Id,
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:833]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 4]
```diff
@@ -0,0 +1,1638 @@
...
+        for (Integer i = 0; i < 5; i++){
+            hed__Education_History__c testEduHist = new hed__Education_History__c(hed__Contact__c = testContact.Id,
+                                                                                  hed__Account__c = firstAccount.Id);
+            eduHistList.add(testEduHist);                                                                      
+        }
+
+        insert eduHistList;
+
+        for (hed__Education_History__c eduHist : eduHistList){
+            Grade_Enrollment__c oldGradeEnroll = new Grade_Enrollment__c(Contact__c = testContact.Id,
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:893]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 5]
```diff
@@ -0,0 +1,1638 @@
...
+        for (Integer i = 0; i < 5; i++){
+            hed__Education_History__c testEduHist = new hed__Education_History__c(hed__Contact__c = testContact.Id,
+                                                                                  hed__Account__c = firstAccount.Id);
+            eduHistList.add(testEduHist);
+        }
+
+        insert eduHistList;
+
+        for (hed__Education_History__c eduHist : eduHistList){
+            Grade_Enrollment__c oldGradeEnroll = new Grade_Enrollment__c(Contact__c = testContact.Id,
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:957]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 6]
```diff
@@ -0,0 +1,1638 @@
...
+        for (Integer i = 0; i < 5; i++){
+            hed__Education_History__c testEduHist = new hed__Education_History__c(hed__Contact__c = testContact.Id,
+                                                                                  hed__Account__c = firstAccount.Id);
+            eduHistList.add(testEduHist);
+        }
+
+        insert eduHistList;
+
+        for (hed__Education_History__c eduHist : eduHistList){
+            Grade_Enrollment__c gradeEnroll = new Grade_Enrollment__c(Contact__c = testContact.Id,
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1023]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 7]
```diff
@@ -0,0 +1,1638 @@
...
+            hed__Education_History__c testEduHist = new hed__Education_History__c(hed__Contact__c = testContact.Id,
+                                                                                  hed__Account__c = firstAccount.Id);
+
+            eduHistList.add(testEduHist);
+        }
+
+        insert eduHistList;
+        
+        for (hed__Education_History__c eduHist : eduHistList){
+            Grade_Enrollment__c gradeEnroll = new Grade_Enrollment__c(Contact__c = testContact.Id,
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1084]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 8]
```diff
@@ -0,0 +1,1638 @@
...
+        
+        for (Integer i = 0; i < 5; i++){
+            Id fakeEduHistId = GRER_DataInteg_TEST.getFakeId(hed__Education_History__c.SObjectType);
+            Id fakeGradeEnrollId = GRER_DataInteg_TEST.getFakeId(Grade_Enrollment__c.SObjectType);
+
+            hed__Education_History__c testEduHist = new hed__Education_History__c(Id = fakeEduHistId,
+                                                                              hed__Contact__c = fakeContactId,
+                                                                              hed__Account__c = fakeFirstAccountId);
+
+            Grade_Enrollment__c gradeEnroll = new Grade_Enrollment__c(Id = fakeGradeEnrollId,
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1133]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 9]
```diff
@@ -0,0 +1,1638 @@
...
+        for (Integer i = 0; i < 5; i++){
+            hed__Education_History__c testEduHist = new hed__Education_History__c(hed__Contact__c = testContact.Id,
+                                                                                  hed__Account__c = firstAccount.Id);
+            eduHistList.add(testEduHist);
+        }
+
+        insert eduHistList;
+
+        for (hed__Education_History__c eduHist : eduHistList){
+            Grade_Enrollment__c gradeEnroll = new Grade_Enrollment__c(Contact__c = testContact.Id,
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1190]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 10]
```diff
@@ -0,0 +1,1638 @@
...
+            hed__Education_History__c testEduHist = new hed__Education_History__c(hed__Contact__c = testContact.Id,
+                                                                                  hed__Account__c = firstAccount.Id);
+            eduHistList.add(testEduHist);
+        }
+
+        insert eduHistList;
+
+        for (hed__Education_History__c eduHist : eduHistList){
+
+            Grade_Enrollment__c oldGradeEnroll = new Grade_Enrollment__c(Contact__c = testContact.Id,
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1249]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 11]
```diff
@@ -0,0 +1,1638 @@
...
+        for (Integer i = 0; i < 5; i++){
+            hed__Education_History__c testEduHist = new hed__Education_History__c(hed__Contact__c = testContact.Id,
+                                                                                  hed__Account__c = firstAccount.Id);
+            eduHistList.add(testEduHist);
+        }
+
+        insert eduHistList;
+
+        for (hed__Education_History__c eduHist : eduHistList){
+            Grade_Enrollment__c oldGradeEnroll = new Grade_Enrollment__c(Contact__c = testContact.Id,
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1314]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 12]
```diff
@@ -0,0 +1,1638 @@
...
+        for (Integer i = 0; i < 5; i++){
+            hed__Education_History__c testEduHist = new hed__Education_History__c(hed__Contact__c = testContact.Id,
+                                                                                  hed__Account__c = firstAccount.Id);
+            eduHistList.add(testEduHist);
+        }
+
+        insert eduHistList;
+
+        for (hed__Education_History__c eduHist : eduHistList){
+            Grade_Enrollment__c gradeEnroll = new Grade_Enrollment__c(Contact__c = testContact.Id,
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1381]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 13]
```diff
@@ -0,0 +1,1638 @@
...
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1443]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 14]
```diff
@@ -0,0 +1,1638 @@
...
+        for (Integer i = 0; i < 5; i++){
+            hed__Education_History__c testEduHist = new hed__Education_History__c(hed__Contact__c = testContact.Id,
+                                                                                  hed__Account__c = firstAccount.Id);
+            eduHistList.add(testEduHist);
+        }
+
+        insert eduHistList;
+
+        for (hed__Education_History__c eduHist : eduHistList){
+            Grade_Enrollment__c oldGradeEnroll = new Grade_Enrollment__c(Contact__c = testContact.Id,
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1503]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 15]
```diff
@@ -0,0 +1,1638 @@
...
+                                                                         Grade_Level__c = 'Eighth Grade',
+                                                                         Education_History__c = eduHist.Id);
+
+            oldGradeEnrollmentsList.add(oldGradeEnroll);
+        }
+
+        insert oldGradeEnrollmentsList;
+
+        for (Grade_Enrollment__c gradeEnroll : oldGradeEnrollmentsList){
+            Grade_Enrollment__c newGradeEnroll = gradeEnroll.clone(true);
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1514]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 16]
```diff
@@ -0,0 +1,1638 @@
...
+            hed__Education_History__c testEduHist = new hed__Education_History__c(hed__Contact__c = testContact.Id,
+                                                                                  hed__Account__c = firstAccount.Id);
+            eduHistList.add(testEduHist);
+        }
+
+        insert eduHistList;
+
+        for (hed__Education_History__c eduHist : eduHistList){
+
+            Grade_Enrollment__c oldGradeEnroll = new Grade_Enrollment__c(Contact__c = testContact.Id,
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1572]: Please use multiple grade enrollments.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 17]
```diff
@@ -0,0 +1,1638 @@
...
+            newGradeEnrollmentsList.add(newGradeEnroll);
+        }
+
+        Test.startTest();
+        GRER_DataInteg_TDTM gradeEnrollTdtm = new GRER_DataInteg_TDTM();
+        List<Exception> exceptionsList = new List<Exception>();
+
+        try{
+            gradeEnrollTdtm.run((List<SObject>)newGradeEnrollmentsList, (List<SObject>)oldGradeEnrollmentsList, triggerAction, objResult);
+        } catch (Exception e){
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1594]: Please catch only for the specific exception type we are expecting.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 18]
```diff
@@ -0,0 +1,1638 @@
...
+        hed.TDTM_Runnable.Action triggerAction = hed.TDTM_Runnable.Action.AfterInsert;
+        Schema.DescribeSObjectResult objResult = Schema.SObjectType.Grade_Enrollment__c;
+
+        Test.startTest();
+        GRER_DataInteg_TDTM gradeEnrollTdtm = new GRER_DataInteg_TDTM();
+        List<Exception> exceptionsList = new List<Exception>();
+
+        try{
+            gradeEnrollTdtm.run(null, null, triggerAction, objResult);
+        } catch (Exception e){
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1618]: Please catch only for the specific exception type we are expecting.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 19]
```diff
@@ -0,0 +1,1638 @@
...
+
+        insert gradeEnrollmentsList;
+        
+        Test.startTest();
+        GRER_DataInteg_TDTM gradeEnrollTdtm = new GRER_DataInteg_TDTM();
+        List<Exception> exceptionsList = new List<Exception>();
+
+        try{
+            gradeEnrollTdtm.checkEducationalInstitutionAgainstEducationHistory((List<SObject>)gradeEnrollmentsList);
+        } catch (Exception e){
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:849]: Please catch only for the specific exception type we are expecting.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 20]
```diff
@@ -0,0 +1,1638 @@
...
+            newGradeEnrollmentsList.add(newGradeEnroll);
+        }
+        
+        Test.startTest();
+        GRER_DataInteg_TDTM gradeEnrollTdtm = new GRER_DataInteg_TDTM();
+        List<Exception> exceptionsList = new List<Exception>();
+
+        try{
+            gradeEnrollTdtm.handleBeforeUpdate((List<SObject>)newGradeEnrollmentsList, (List<SObject>)oldGradeEnrollmentsList);
+        } catch (Exception e){
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:915]: Please catch only for the specific exception type we are expecting.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 21]
```diff
@@ -0,0 +1,1638 @@
...
+            newGradeEnrollmentsList.add(newGradeEnroll);
+        }
+
+        Test.startTest();
+        GRER_DataInteg_TDTM gradeEnrollTdtm = new GRER_DataInteg_TDTM();
+        List<Exception> exceptionsList = new List<Exception>();
+
+        try{
+            gradeEnrollTdtm.handleBeforeUpdate((List<SObject>)newGradeEnrollmentsList, (List<SObject>)oldGradeEnrollmentsList);
+        } catch (Exception e){
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:980]: Please catch only for the specific exception type we are expecting.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 22]
```diff
@@ -0,0 +1,1638 @@
...
+
+        insert gradeEnrollmentsList;
+
+        GRER_DataInteg_TDTM gradeEnrollTdtm = new GRER_DataInteg_TDTM();
+        List<Exception> exceptionsList = new List<Exception>();
+
+        Test.startTest();
+        try{
+            gradeEnrollTdtm.handleAfterInsert((List<SObject>)gradeEnrollmentsList);
+        } catch (Exception e){
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1040]: Please catch only for the specific exception type we are expecting.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 23]
```diff
@@ -0,0 +1,1638 @@
...
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1101]: Please catch only for the specific exception type we are expecting.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 24]
```diff
@@ -0,0 +1,1638 @@
...
+            gradeEnrollmentsList.add(gradeEnroll);
+        }
+
+        GRER_DataInteg_TDTM gradeEnrollTdtm = new GRER_DataInteg_TDTM();
+        List<Exception> exceptionsList = new List<Exception>();
+
+        Test.startTest();
+        try{
+            gradeEnrollTdtm.handleAfterInsert((List<SObject>)gradeEnrollmentsList);
+        } catch (Exception e){
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1149]: Please catch only for the specific exception type we are expecting.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 25]
```diff
@@ -0,0 +1,1638 @@
...
+
+        insert gradeEnrollmentsList;
+
+        GRER_DataInteg_TDTM gradeEnrollTdtm = new GRER_DataInteg_TDTM();
+        List<Exception> exceptionsList = new List<Exception>();
+
+        Test.startTest();
+        try{
+            gradeEnrollTdtm.checkEducationalInstitutionAgainstEducationHistory((List<SObject>)gradeEnrollmentsList);
+        } catch (Exception e){
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1207]: Please catch only for the specific exception type we are expecting.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 26]
```diff
@@ -0,0 +1,1638 @@
...
+            newGradeEnrollmentsList.add(newGradeEnroll);
+        }
+
+        Test.startTest();
+        GRER_DataInteg_TDTM gradeEnrollTdtm = new GRER_DataInteg_TDTM();
+        List<Exception> exceptionsList = new List<Exception>();
+
+        try{
+            gradeEnrollTdtm.checkEducationalInstitutionAgainstEducationHistory((List<SObject>)newGradeEnrollmentsList, (List<SObject>)oldGradeEnrollmentsList);
+        } catch (Exception e){
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1273]: Please catch only for the specific exception type we are expecting.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 27]
```diff
@@ -0,0 +1,1638 @@
...
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1336]: Please catch only for the specific exception type we are expecting.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 28]
```diff
@@ -0,0 +1,1638 @@
...
+
+        insert gradeEnrollmentsList;
+
+        Test.startTest();
+        GRER_DataInteg_TDTM gradeEnrollTdtm = new GRER_DataInteg_TDTM();
+        List<Exception> exceptionsList = new List<Exception>();
+
+        try{
+            gradeEnrollTdtm.run((List<SObject>)gradeEnrollmentsList, null, triggerAction, objResult);
+        } catch (Exception e){
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1398]: Please catch only for the specific exception type we are expecting.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 29]
```diff
@@ -0,0 +1,1638 @@
...
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1459]: Please catch only for the specific exception type we are expecting.

File: src/classes/GRER_DataInteg_TEST.cls [hunk 30]
```diff
@@ -0,0 +1,1638 @@
...
+            newGradeEnrollmentsList.add(newGradeEnroll);
+        }
+
+        GRER_DataInteg_TDTM gradeEnrollTdtm = new GRER_DataInteg_TDTM();
+        List<Exception> exceptionsList = new List<Exception>();
+
+        Test.startTest();
+        try{
+            gradeEnrollTdtm.run((List<SObject>)newGradeEnrollmentsList, (List<SObject>)oldGradeEnrollmentsList, triggerAction, objResult);
+        } catch (Exception e){
```
Comment (MatthewBlanski) [inline src/classes/GRER_DataInteg_TEST.cls:1525]: Please catch only for the specific exception type we are expecting.

**Review by MatthewBlanski — CHANGES_REQUESTED** [review]
There are a number of changes requested. The most salient is handling multiple dependent records and some logic flow changes within the specific methods to reduce the number of loops and query only relevant records.

There are a few specific test pieces to handle as well, but otherwise the code is clean and effective.

**Review by MatthewBlanski — APPROVED** [review]
This is much improved and targets relevant records only as well as uses multiple records in its unit tests.

**Comment by spinkelman** [comment]
Passes QA - Merging
Will Close WI

**Pull request merged**